### PR TITLE
Include paths #54

### PR DESCRIPTION
**Pull request by YuraYelisieiev**
Fixes #	

 Changes:	
-	

 Does this change need to mentioned in CHANGELOG.md?	
no	

 Requires testing	
no

 Comments about testing, should you have some

### CODE REVIEW

**Comment by stdevAlDen** [comment]
@YuraYelisieiev please make fetch and merge from upstream/develop and solve the conflicts

**Comment by stdevAlDen** [comment]
> @YuraYelisieiev please make fetch and merge from upstream/develop and solve the conflicts

@YuraYelisieiev 
you should use the pr template, describe in a short item what are the changes for and put a reference to the original issue to close it
https://help.github.com/en/articles/closing-issues-using-keywords

**Comment by stdevAlDen** [comment]
@YuraYelisieiev we can not call `my_banch` because this not give much information look [at this](https://github.com/skycoin/hardware-wallet/pull/198) the branch name follow the pattern
YourUserName_tIsueId_ANameFromTheTittleButShorter
and in snake case, not cammel case... look at the example, it's just a good practice

**Comment by YuraYelisieiev** [comment]
@stdevAlDen fixed all the issues now. Waiting for your review.

File: tiny-firmware/Makefile.include [hunk 1]
```diff
@@ -110,11 +110,10 @@ CFLAGS   += $(OPTFLAGS) \
             -DSTM32F2 \
             -DRAND_PLATFORM_INDEPENDENT=1 \
             -I$(TOOLCHAIN_DIR)/include \
-            -I$(TOP_DIR) \
             -I$(TOP_DIR)gen \
-            -I$(TOP_DIR)vendor/skycoin-crypto \
-            -I$(TOP_DIR)vendor/skycoin-crypto/tools \
-            -I$(TOP_DIR)vendor/skycoin-crypto/tools/aes
+            -isystem$(TOP_DIR)vendor/skycoin-crypto \
+            -isystem$(TOP_DIR)vendor/skycoin-crypto/tools \
```
Comment (stdevAlDen) [inline tiny-firmware/Makefile.include]: @YuraYelisieiev plese only one to `-isystem$(TOP_DIR)vendor`

File: tiny-firmware/bootloader/signatures.c [hunk 2]
```diff
@@ -21,11 +21,13 @@
 #include <stdint.h>
 #include <string.h>
 
-#include "bootloader.h"
 #include "sha2.h"
 #include "signatures.h"
+
+#include "bootloader.h"
```
Comment (stdevAlDen) [inline tiny-firmware/bootloader/signatures.c]: @YuraYelisieiev why this works?
`tiny-firmware/bootloader/` still in the include search path.

File: tiny-firmware/bootloader/usb.c [hunk 3]
```diff
@@ -24,19 +24,21 @@
 
 #include <string.h>
 
+#include "usb.h"
+
 #include "bootloader.h"
-#include "buttons.h"
+#include "signatures.h"
```
Comment (stdevAlDen) [inline tiny-firmware/bootloader/usb.c]: @YuraYelisieiev the same as the above

File: tiny-firmware/firmware/fsm_impl.c [hunk 4]
```diff
@@ -24,23 +25,25 @@
 #include "fsm_impl.h"
 #include "gettext.h"
 #include "layout2.h"
-#include "memory.h"
 #include "messages.h"
-#include "oled.h"
-#include "pinmatrix.h"
+#include "storage.h"
```
Comment (stdevAlDen) [inline tiny-firmware/firmware/fsm_impl.c]: @YuraYelisieiev full path please

File: tiny-firmware/Makefile [hunk 5]
```diff
@@ -43,13 +43,15 @@ endif
 
 DEBUG_LINK ?= 0
 CFLAGS += -Wno-sequence-point
+CFLAGS += -I$(TOP_DIR)../
```
Comment (olemis) [inline tiny-firmware/Makefile]: Is there any `.h` file at this location ? Is this necessary ?

Comment (stdevAlDen) [reply tiny-firmware/Makefile]: @olemis the issue purpose is to remove those folders where `*.h` files found from the include path and just add the root dir, after this all the `#indlude ...` should look like(for example):
```.c
#include "tiny-firmware/layout.h"
#include "tiny-firmware/firmware/droplet.h"
#include <libopencm3/usb/cdc.h>
#include <libopencm3/usb/usbd.h>
```

Question: for the crypto API what do you recommend?
`#include "skycoin-api/skycoin_crypto.h"`
or
`#include <skycoin-crypto/skycoin_crypto.h>
I means, from vendor(trough a link) or from root dir? @olemis what do you think on this?

Comment (stdevAlDen) [inline tiny-firmware/Makefile]: @YuraYelisieiev please replace the `..` shortcut with the [root dir var](https://github.com/skycoin/hardware-wallet/blob/c542b06eae1516896052284f3767abba350fae5d/Makefile#L18).

**Comment by YuraYelisieiev** [comment]
@stdevAlDen Take a look on my fixes. I still need to include only the vendor folder.

File: tiny-firmware/timer.c [hunk 6]
```diff
@@ -18,14 +18,14 @@
...
+#include "tiny-firmware/timer.h"
 
 #include <libopencm3/cm3/systick.h>
 #include <libopencm3/cm3/vector.h>
 #include <libopencm3/stm32/rcc.h>
 
-#include "firmware/swtimer.h"
-#include "rng.h"
+#include <tiny-firmware/rng.h>
+#include <tiny-firmware/firmware/swtimer.h>
```
Comment (stdevAlDen) [inline tiny-firmware/timer.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/swtimer.h"`

File: tiny-firmware/firmware/main.c [hunk 7]
```diff
@@ -11,23 +11,23 @@
...
+#include <tiny-firmware/layout.h>
+#include <tiny-firmware/firmware/layout2.h>
+#include <tiny-firmware/rng.h>
+#include <tiny-firmware/timer.h>
+#include <tiny-firmware/buttons.h>
+#include <tiny-firmware/firmware/gettext.h>
+#include <tiny-firmware/firmware/fastflash.h>
+#include <tiny-firmware/firmware/factory_test.h>
+#include <tiny-firmware/firmware/entropy.h>
+#include <tiny-firmware/memory.h>
```
Comment (stdevAlDen) [inline tiny-firmware/firmware/main.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/*"`

File: tiny-firmware/firmware/layout2.c [hunk 8]
```diff
@@ -23,15 +23,17 @@
...
-#include "bitmaps.h"
-#include "gettext.h"
-#include "layout2.h"
-#include "oled.h"
-#include "secp256k1.h"
-#include "storage.h"
+#include <tiny-firmware/firmware/gettext.h>
+#include <tiny-firmware/firmware/layout2.h>
+#include <tiny-firmware/oled.h>
+#include <tiny-firmware/gen/bitmaps.h>
```
Comment (stdevAlDen) [inline tiny-firmware/firmware/layout2.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/*"`

File: tiny-firmware/firmware/fsm_impl.c [hunk 9]
```diff
@@ -9,38 +9,35 @@
...
+#include <tiny-firmware/firmware/droplet.h>
+#include <tiny-firmware/firmware/entropy.h>
+#include <tiny-firmware/firmware/fsm.h>
+#include <tiny-firmware/firmware/fsm_impl.h>
+#include <tiny-firmware/firmware/gettext.h>
+#include <tiny-firmware/firmware/layout2.h>
+#include <tiny-firmware/firmware/messages.h>
+#include <tiny-firmware/firmware/storage.h>
+#include <tiny-firmware/rng.h>
+#include <tiny-firmware/oled.h>
```
Comment (stdevAlDen) [inline tiny-firmware/firmware/fsm_impl.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/*"`

File: tiny-firmware/firmware/factory_test.c [hunk 10]
```diff
@@ -11,17 +11,17 @@
...
 #else
-#include "timer.h"
+#include <tiny-firmware/timer.h>
 #endif
-#include "buttons.h"
-#include "layout.h"
-#include "layout2.h"
-#include "util.h"
+#include <tiny-firmware/buttons.h>
+#include <tiny-firmware/layout.h>
```
Comment (stdevAlDen) [inline tiny-firmware/firmware/factory_test.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/*"`

File: tiny-firmware/firmware/entropy.c [hunk 11]
```diff
@@ -18,17 +18,15 @@
 #include "gpio_noise.h"
 
 #endif // EMULATOR
-
-#include "firmware/storage.h"
-#include "firmware/swtimer.h"
-#include "messages.h"
+#include <tiny-firmware/rng.h>
+#include <tiny-firmware/firmware/swtimer.h>
```
Comment (stdevAlDen) [inline tiny-firmware/firmware/entropy.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/*"`

File: tiny-firmware/emulator/setup.c [hunk 12]
```diff
@@ -26,11 +26,11 @@
 
 #include <libopencm3/stm32/flash.h>
 
-#include "memory.h"
-#include "oled.h"
-#include "rng.h"
-#include "setup.h"
-#include "timer.h"
+#include <tiny-firmware/memory.h>
+#include <tiny-firmware/oled.h>
```
Comment (stdevAlDen) [inline tiny-firmware/emulator/setup.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/*"`

File: tiny-firmware/bootloader/bootloader.c [hunk 13]
```diff
@@ -24,17 +24,17 @@
...
-#include "usb.h"
-#include "util.h"
-
+#include <tiny-firmware/bootloader/bootloader.h>
+#include <tiny-firmware/bootloader/signatures.h>
+#include <tiny-firmware/bootloader/usb.h>
+
+#include <tiny-firmware/buttons.h>
+#include <tiny-firmware/setup.h>
+#include <tiny-firmware/oled.h>
```
Comment (stdevAlDen) [inline tiny-firmware/bootloader/bootloader.c]: @YuraYelisieiev please change to `"tiny-firmware/firmware/*"`

File: tiny-firmware/Makefile.include [hunk 14]
```diff
@@ -110,11 +110,11 @@ CFLAGS   += $(OPTFLAGS) \
             -DSTM32F2 \
             -DRAND_PLATFORM_INDEPENDENT=1 \
             -I$(TOOLCHAIN_DIR)/include \
-            -I$(TOP_DIR) \
             -I$(TOP_DIR)gen \
-            -I$(TOP_DIR)vendor/skycoin-crypto \
-            -I$(TOP_DIR)vendor/skycoin-crypto/tools \
-            -I$(TOP_DIR)vendor/skycoin-crypto/tools/aes
+            -isystem$(TOP_DIR)vendor/skycoin-crypto \
```
Comment (stdevAlDen) [inline tiny-firmware/Makefile.include]: @YuraYelisieiev a change from the original request, due to [this](https://github.com/skycoin/hardware-wallet/pull/254) we now need  to remove the:
`-isystem$(TOP_DIR)vendor/skycoin-crypto` and `-isystem$(TOP_DIR)vendor/skycoin-crypto/tools/aes`. So we need to include with full path, starting from `GLOBAL_PATH` like `#include "skycoin-api/tools/bip32.h"` and remove this symlink (tiny-firmware/vendor/skycoin-crypto)

**Comment by YuraYelisieiev** [comment]
@stdevAlDen If we will change to double quotes it will start to search from the current directory. So why we choose it ?

**Comment by YuraYelisieiev** [comment]
@stdevAlDen I changed only some of them because wanted to discuss it with you, how to do it in a better way.

**Comment by stdevAlDen** [comment]
> @stdevAlDen I changed only some of them because wanted to discuss it with you, how to do it in a better way.

great @YuraYelisieiev I'm open to discuss with you about this, to take the best possible decision together, but sorry can you please get me more details about your opinion, I can not understand your proposal if any.

**Comment by YuraYelisieiev** [comment]
@stdevAlDen If we use include`s like this ```#include <tiny-firmware/buttons.h>``` then it starts the search from the root directory. But when we use ```#include \"tiny-firmware/buttons.h\"``` it starts from the current, iterates through every subdirectory and the starts with the directories which we included with the -I    flag.

**Comment by YuraYelisieiev** [comment]
@stdevAlDen Also if we don`t want developers to include files from the current directory, we need to force them to use the first version, because with double quotes you can still include files without the absolute path.

**Comment by stdevAlDen** [comment]
> @stdevAlDen If we use include`s like this `#include <tiny-firmware/buttons.h>` then it starts the search from the root directory. But when we use ```#include "tiny-firmware/buttons.h"``` it starts from the current, iterates through every subdirectory and the starts with the directories which we included with the -I flag.

@YuraYelisieiev use `<>` for system and 3rd party libraries, `""` for the project sources.

**Comment by stdevAlDen** [comment]
> @stdevAlDen Also if we don`t want developers to include files from the current directory, we need to force them to use the first version, because with double quotes you can still include files without the absolute path.

@YuraYelisieiev please note the difference between compiler and a project convention. Do not worry to much if the compiler can not catch this as an error, in any case this is valid C, so to prevent this we can include a linter stage in the CI.
What is really important (from my point of view) is [...](https://github.com/skycoin/hardware-wallet/pull/227#issuecomment-513943657)

**Comment by YuraYelisieiev** [comment]
@stdevAlDen changed it, review pls.

File: skycoin-api/skycoin_crypto.c [hunk 15]
```diff
@@ -16,12 +16,12 @@
...
-#include "curves.h"
-#include "ecdsa.h"
-#include "ripemd160.h"
-#include "sha2.h"
+#include <tools/base58.h>
+#include <tools/secp256k1.h>
+#include <tools/curves.h>
+#include <tools/ecdsa.h>
+#include <tools/ripemd160.h>
+#include <tools/sha2.h>
```
Comment (stdevAlDen) [inline skycoin-api/skycoin_crypto.c]: @YuraYelisieiev as I mention [here](https://github.com/skycoin/hardware-wallet/pull/227#discussion_r297790263), the crypto API is not considered 3rd party any more, so this is par of out project.

Comment (YuraYelisieiev) [reply skycoin-api/skycoin_crypto.c]: @stdevAlDen fixed.

**Comment by jdknives** [comment]
@stdevAlDen would be appreciated if you could give this a final review so it can be merged.

**Comment by YuraYelisieiev** [comment]
@stdevAlDen missed one PR wich was merged into develop with Makefile change. Will make travis pass in a while

**Comment by YuraYelisieiev** [comment]
@stdevAlDen done now. Review the changes pls.

**Comment by stdevAlDen** [comment]
> @stdevAlDen done now. Review the changes pls.

@YuraYelisieiev 
The rules should be easy to apply, for example [here](https://github.com/skycoin/hardware-wallet/issues/54#issue-400911432), [here](https://github.com/skycoin/hardware-wallet/pull/227#issuecomment-513943657), and maybe some other comments. With a first look I can find some remaining issues, for example [here](https://github.com/skycoin/hardware-wallet/pull/227/files#diff-e0336b1d462466865833223db846e4acR21).

@jdknives I think that this issue is not important at all for this milestone, even I like to avoid this merge for this milestone. I assigned the next milestone, so fix it if I'm wrong and let us know about any final decision on this.

**Comment by YuraYelisieiev** [comment]
@stdevAlDen probably missing something cant see what is wrong [here](https://github.com/skycoin/hardware-wallet/issues/54#issue-400911432)

File: tiny-firmware/Makefile.include (hunk 14 above)
Comment (RomanMilishchuk) [reply tiny-firmware/Makefile.include]: Just change -isystem$(TOP_DIR)vendor/skycoin-crypto to -isystem$(TOP_DIR)vendor/

File: tiny-firmware/bootloader/signatures.c [hunk 16]
```diff
@@ -21,9 +21,9 @@
 #include <stdint.h>
 #include <string.h>
 
-#include "bootloader.h"
-#include "sha2.h"
-#include "signatures.h"
+#include "tools/sha2.h"
```
Comment (RomanMilishchuk) [inline tiny-firmware/bootloader/signatures.c]: So then here will be skycoin-crypto/tools/sha2.h and we will see in what library these headers are

**Pull request merged**